src/group_theory/divisible.lean

```diff
@@ -0,0 +1,264 @@
+    generalize_proofs h1,
+    exact h1.some_spec.2,
+  end }
+
+@[priority 100]
+instance divisible_of_char_zero {𝕜} [division_ring 𝕜] [char_zero 𝕜] : divisible_by 𝕜 ℤ :=
+{ div := λ q n, q / n,
+  div_zero := λ q, by norm_num,
+  div_cancel := λ n q hn,
+    by rw [zsmul_eq_mul, (int.cast_commute n _).eq, div_mul_cancel q (int.cast_ne_zero.mpr hn)] }
+
+end add_comm_group
+
+namespace group
+
+open monoid
+
+variables (A : Type*) [group A]
+
+/--
+A group is `ℤ`-rootable if it is `ℕ`-rootable.
+-/
+@[to_additive add_group.divisible_by_int_of_divisible_by_nat
+"An add group is `ℤ`-divisible if it is `ℕ`-divisible."]
+def rootable_by_int_of_rootable_by_nat [rootable_by A ℕ] :
+  rootable_by A ℤ :=
+{ root := λ a z, match z with
+  | int.of_nat n := rootable_by.root a n
+  | int.neg_succ_of_nat n := (rootable_by.root a (n + 1))⁻¹
+  end,
+  root_zero := λ a, rootable_by.root_zero a,
+  root_cancel := λ n a hn, begin
+    induction n,
+    { change (rootable_by.root a _) ^ _ = a,
+      norm_num,
+      rw [rootable_by.root_cancel],
+      rw [int.of_nat_eq_coe] at hn,
+      exact_mod_cast hn, },
+    { change ((rootable_by.root a _) ⁻¹)^_ = a,
+      norm_num,
+      rw [rootable_by.root_cancel],
+      norm_num, }
+  end}
+
+/--A group is `ℕ`-rootable if it is `ℤ`-rootable
+-/
+@[to_additive add_group.divisible_by_nat_of_divisible_by_int
+"An add group is `ℕ`-divisible if it `ℤ`-divisible."]
+def rootable_by_nat_of_rootable_by_int [rootable_by A ℤ] :
+  rootable_by A ℕ :=
+{ root := λ a n, rootable_by.root a (n : ℤ),
+  root_zero := λ a, rootable_by.root_zero a,
+  root_cancel := λ n a hn, begin
+    have := rootable_by.root_cancel a (show (n : ℤ) ≠ 0, by exact_mod_cast hn),
+    norm_num at this,
+    exact this,
+  end }
+
+end group
+
+namespace comm_group
+
+open monoid
+
+section quotient
+
+variables (A : Type*) [comm_group A] (B : subgroup A)
+
+/--
+Any quotient group of a rootable group is rootable.
+-/
+@[to_additive add_comm_group.divisible_by_quotient
+"Any quotient group of a divisible group is divisible"]
+noncomputable def rootable_by_quotient [rootable_by A ℕ] : rootable_by (A ⧸ B) ℕ :=
+rootable_by_of_pow_surj _ _ $ λ n hn x, quotient.induction_on' x $ λ a,
+  ⟨quotient.mk' (rootable_by.root a n),
+    (congr_arg _ $ rootable_by.root_cancel _ hn : quotient.mk' _ = _)⟩
+
+end quotient
+
+section hom
+
+variables {A B : Type*} [comm_group A] [comm_group B] [rootable_by A ℕ] (f : A →* B)
+
+/--
+If `f : A → B` is a surjective homomorphism and `A` is rootable, then `B` is also rootable.
+-/
+@[to_additive add_comm_group.divisible_by_of_surj
+"If `f : A → B` is a surjective homomorphism and `A` is divisble, then `B` is also divisible."]
+noncomputable def rootable_by_of_surj (hf : function.surjective f) : rootable_by B ℕ :=
+rootable_by_of_pow_surj _ _ $ λ n hn x,
+  ⟨f $ rootable_by.root (hf x).some n, (by rw [←f.map_pow (rootable_by.root (hf x).some n) n,
+    rootable_by.root_cancel _ hn, (hf x).some_spec] : _ ^ _ = x)⟩
```

Comment:
```suggestion
rootable_by_of_pow_surj _ _ $ λ n hn x,
  let ⟨y, hy⟩ := hf x in ⟨f $ rootable_by.root y n, (by rw [←f.map_pow (rootable_by.root y n) n,
    rootable_by.root_cancel _ hn, hy] : _ ^ _ = x)⟩
```
No need for `some` here

Reply:
Didn't know that you can destruct inside `let`